dag_in_context/src/pretty_print.rs

```diff
@@ -0,0 +1,831 @@
+use crate::{
+    from_egglog::FromEgglog,
+    prologue,
+    schema::{
+        self, Assumption, BaseType, BinaryOp, Expr, RcExpr, TernaryOp, TreeProgram, Type, UnaryOp,
+    },
+    schema_helpers::AssumptionRef,
+    to_egglog::TreeToEgglog,
+};
+use egglog::{Term, TermDag};
+
+use std::{collections::HashMap, hash::Hash, rc::Rc, vec};
+
+pub struct PrettyPrinter {
+    pub expr: RcExpr,
+    // Type/Assum/BaseType -> intermediate variables
+    symbols: indexmap::IndexMap<Symbols, String>,
+    // intermediate variables about to print
+    log: Vec<String>,
+    // intermediate variable -> Type/Assum/BaseType lookup
+    table: std::collections::BTreeMap<String, SymbolLog>,
```

Comment:
Why not HashMap here?

Reply:
because Symbolog contain Expr and Assumption which cannot be hashed